controllers/operatorpolicy_controller.go

```diff
@@ -60,166 +70,211 @@ var _ reconcile.Reconciler = &OperatorPolicyReconciler{}
 			if err != nil {
+				OpLog.Error(err, "Error while creating OperatorGroup")
 				r.setCompliance(policy, policyv1.NonCompliant)
-				OpLog.Error(err, "Could not handle missing musthave object")
 
-				return err
+				err = r.DynamicWatcher.EndQueryBatch(watcher)
+				if err != nil {
+					panic(err)
+				}
+
+				return reconcile.Result{}, err
 			}
 
-			// Currently creates an OperatorGroup for every Subscription
-			// in the same ns, and defaults to targeting all ns.
-			// Future implementations will enable targeting ns based on
-			// installModes supported by the CSV. Also, only one OperatorGroup
-			// should exist in each ns
-			operatorGroup := buildOperatorGroup(policy)
-			err = r.Create(context.TODO(), operatorGroup)
+			// Created successfully, requeue result
+			r.setCompliance(policy, policyv1.NonCompliant)
 
+			err = r.DynamicWatcher.EndQueryBatch(watcher)
 			if err != nil {
-				r.setCompliance(policy, policyv1.NonCompliant)
-				OpLog.Error(err, "Could not handle missing musthave object")
-
-				return err
+				panic(err)
 			}
 
-			r.setCompliance(policy, policyv1.Compliant)
+			return reconcile.Result{Requeue: true}, err
+		} else if policy.Spec.OperatorGroup != nil {
+			// If inform mode, keep going and return at the end without requeue. Before
+			// continuing, should check if required og spec is created to set compliance state
 
-			return nil
+			// Set to non compliant because operatorgroup does not exist on cluster, but
+			// should still go to check subscription
+			r.setCompliance(policy, policyv1.NonCompliant)
 		}
+	}
 
-		// Inform
-		r.setCompliance(policy, policyv1.NonCompliant)
-
-		return nil
+	// Retrieve subscription from cache
+	cachedSubscription, err := r.DynamicWatcher.GetFromCache(
+		subscriptiongvk,
+		policy.Spec.Subscription.Namespace, policy.Spec.Subscription.SubscriptionSpec.Package)
+	if err != nil {
+		panic(err)
 	}
 
-	// Object exists but it should not exist, delete object
-	// Deleting related objects will be added in the future
-	if exists && !shouldExist {
+	// Create new subscription
+	if cachedSubscription == nil {
 		if strings.EqualFold(string(policy.Spec.RemediationAction), string(policyv1.Enforce)) {
-			OpLog.Info("deleting kind " + subscriptionSpec.Kind + " in ns " + subscriptionSpec.Namespace)
-			err = r.Delete(context.TODO(), subscriptionSpec)
+			subscriptionSpec := buildSubscription(policy)
+			OpLog.Info("creating kind " + subscriptionSpec.Kind + " in ns " + subscriptionSpec.Namespace)
 
+			err := r.Create(context.TODO(), subscriptionSpec)
 			if err != nil {
 				r.setCompliance(policy, policyv1.NonCompliant)
-				OpLog.Error(err, "Could not handle existing musthave object")
+				OpLog.Error(err, "Could not handle missing musthave object")
+
+				err = r.DynamicWatcher.EndQueryBatch(watcher)
+				if err != nil {
+					panic(err)
+				}
 
-				return err
+				return reconcile.Result{}, err
 			}
 
 			r.setCompliance(policy, policyv1.Compliant)
```

Comment:
Is this actually compliant? Don't you need to wait for the status checks to confirm it's compliant?

Reply:
Currently, as long as the controller is able to create the necessary resources specified in the OperatorPolicy, it will be considered 'Complaint'. In future PRs, specifically after adding the health checks for the deployments, the compliance state of the policy should be more definitive and better represented with respect to the state of the resources it will watch. Hope that makes sense!

Reply:
Sure, that's fine. Could you please add a comment in the code to that effect?